sparse/mlir_backend/levels.py

```diff
@@ -108,15 +109,28 @@ def _get_ctypes_type(self, *, owns_memory=False):
         def get_fields():
             fields = []
             compressed_counter = 0
+            singleton_counter = 0
             for level, next_level in itertools.zip_longest(self.levels, self.levels[1:]):
                 if LevelFormat.Compressed == level.format:
                     compressed_counter += 1
                     fields.append((f"pointers_to_{compressed_counter}", get_nd_memref_descr(1, ptr_dtype)))
                     if next_level is not None and LevelFormat.Singleton == next_level.format:
-                        fields.append((f"indices_{compressed_counter}", get_nd_memref_descr(2, idx_dtype)))
+                        singleton_counter += 1
+                        fields.append(
+                            (
+                                f"indices_{compressed_counter}_coords_{singleton_counter}",
+                                get_nd_memref_descr(1, idx_dtype),
+                            )
+                        )
                     else:
                         fields.append((f"indices_{compressed_counter}", get_nd_memref_descr(1, idx_dtype)))
 
+                if LevelFormat.Singleton == level.format:
+                    singleton_counter += 1
+                    fields.append(
+                        (f"indices_{compressed_counter}_coords_{singleton_counter}", get_nd_memref_descr(1, idx_dtype))
+                    )
+
```

Comment:
This should probably handle SOA and without SOA separately.

Reply:
In my opinion we should only support SoA singleton format:
1. Non-SoA singleton looks to be buggy for basic operations [link](https://discourse.llvm.org/t/passmanager-fails-on-simple-coo-addition-example/81247/2?u=mtsokol)
2. Mixed singleton levels aren't allowed: https://github.com/llvm/llvm-project/blob/8d38fbf2f027c72332c8ba03ff0ff0f83b4dcf02/mlir/lib/Dialect/SparseTensor/IR/SparseTensorDialect.cpp#L811

What would be the benefit of supporting non-SoA singleton levels separately?

Reply:
We'd be able to support the current COO format only for non-SoA.

Reply:
What do you mean by that? Can you give some example?
With this PR we can support COO format (also input object from `scipy.sparse.coo_array`) where MLIR-backend implementation uses SoA representation.

Reply:
The existing Numba `COO` format uses the non-SoA format, and we pretty much have to support this to be backwards compatible.

Reply:
I think I'm still missing the point here. Numba is a separate backend that supports only 1D and 2D COO arrays. MLIR backend supports >=2D COO arrays.
What do you mean by backward compatibility here? Can you give an example where backward compatibility breaks here? If a user passes `scipy.sparse.coo_array` object to `sparse.asarray` function it it will work for any backend regardless of an internal representation (SoA or non-SoA).

Reply:
So, `sparse.COO` has a constructor that takes `coords` and a `.coords` attribute. The attribute is a 2D NumPy array.

Ideally, I'd like to keep it a 2D `np.ndarray` as otherwise I'm not 100% sure how much will break.

We could do this with `np.stack`, but that would incur a performance penalty.

Reply:
I think for accessing `coords`, `np.stack` could be an option. My caveat for non-SoA in MLIR backend is that it already has some issues with basic operations: https://discourse.llvm.org/t/passmanager-fails-on-simple-coo-addition-example/81247